cmd/util/ledger/util/payload_snapshot.go

```diff
@@ -123,27 +126,139 @@ func (p PayloadSnapshot) ApplyChangesAndGetNewPayloads(
+func NewMapBasedPayloadSnapshotWithWorkers(log zerolog.Logger, payloads []*ledger.Payload, workers int) (*MapBasedPayloadSnapshot, error) {
+
+	// limit the number of workers to prevent too many small jobs
+	const minPayloadsPerWorker = 10000
+	if len(payloads)/minPayloadsPerWorker < workers {
+		workers = len(payloads) / minPayloadsPerWorker
+	}
+	if workers < 1 {
+		workers = 1
+	}
+
+	// log the creation of the snapshot if it is large
+	if len(payloads) > minPayloadsPerWorker {
+		start := time.Now()
+		log.Info().
+			Int("payloads", len(payloads)).
+			Int("workers", workers).
+			Msgf("Creating payload snapshot")
+
+		defer func() {
+			log.Info().
+				Int("payloads", len(payloads)).
+				Int("workers", workers).
+				Dur("duration", time.Since(start)).
+				Msgf("Created payload snapshot")
+		}()
+	}
+
 	payloadsCopy := make([]*ledger.Payload, len(payloads))
 	copy(payloadsCopy, payloads)
-	l := &MapBasedPayloadSnapshot{
-		reverseMap: make(map[flow.RegisterID]int, len(payloads)),
-		payloads:   payloadsCopy,
+
+	type result struct {
+		minimap map[flow.RegisterID]int
+		err     error
 	}
-	for i, payload := range payloadsCopy {
-		key, err := payload.Key()
-		if err != nil {
-			return nil, err
+
+	minimaps := make(chan result, workers)
+	defer close(minimaps)
+
+	// create the minimaps in parallel
+	for i := 0; i < workers; i++ {
+		start := i * len(payloadsCopy) / workers
+		end := (i + 1) * len(payloadsCopy) / workers
+		if end > len(payloadsCopy) {
+			end = len(payloadsCopy)
 		}
-		id, err := convert.LedgerKeyToRegisterID(key)
-		if err != nil {
-			return nil, err
+		go func(payloads []*ledger.Payload) {
+			minimap := make(map[flow.RegisterID]int, len(payloads))
+			for i, payload := range payloads {
+				key, err := payload.Key()
+				if err != nil {
+					minimaps <- result{nil, err}
+					return
+				}
+				id, err := convert.LedgerKeyToRegisterID(key)
+				if err != nil {
+					minimaps <- result{nil, err}
+					return
+				}
+				minimap[id] = i
+			}
+			minimaps <- result{minimap, nil}
+		}(payloadsCopy[start:end])
+	}
+
+	// merge the minimaps in parallel
+	pairedMinimaps := make(
+		chan struct {
+			left  result
+			right result
+		}, workers/2)
+
+	go func() {
+		// we have to pair the minimaps to merge them
+		// we have to merge a total of workers-1 times
+		// before we are left with only one map
+		numberOfPairings := workers - 1
+		for i := 0; i < numberOfPairings; i++ {
+			left := <-minimaps
+			right := <-minimaps
+			pairedMinimaps <- struct {
+				left  result
+				right result
+			}{left, right}
 		}
-		l.reverseMap[id] = i
+		close(pairedMinimaps)
+	}()
+
+	// only half of the workers are needed to merge the maps
+	wg := sync.WaitGroup{}
+	wg.Add(workers / 2)
+	for i := 0; i < workers/2; i++ {
+		go func() {
+			defer wg.Done()
+			for pair := range pairedMinimaps {
+				if pair.left.err != nil {
+					minimaps <- result{nil, pair.left.err}
+					return
```

Comment:
I believe if this returns the `pairedMinimaps` will be blocked to be written to since no one will read it to release the buffer. I guess not critical in this use, but not sure so rather write it.